blog/zk-snark-pallet.md

```diff
@@ -0,0 +1,263 @@
+# Zk-SNARKs with Substrate (Part 3)
+
+This is the last part of a series of articles about zk-SNARKs. If you haven't read the previous ones (Part 1, Part 2), I encourage you to do so now.
+
+In our first article, we defined Bob's problem. He creates a contest, where the first person who solves the equation: 
+
+<center>
+
+$$ x^2+3=12 $$
+
+</center>
+
+will be able to join his Bright Coders union. As you probably remember, Alice was one of his friends who knows the solution. She was only afraid that revealing it loudly, could encourage others to first claim the vacancy. That's why she decided to use zk-SNARKs.
+
+Until this moment, everything that we did assumed Bob and Alice are in the same place. Alice decides to create proof because she doesn't want to reveal the solution to anyone. Bob verified it in front of her, so Alice was certain that she was the first who solve the puzzle. But what if they are not in the same place? What guarantee will have Alice, that her solution was verified first? This problem can be easily solved, if we could move it to the blockchain! Information about the winner will be known to everyone, and the verification process will be more transparent.
+
+We are going to use a framework called [Substrate](https://substrate.io/), to create a custom blockchain. It is written in Rust language and was created by the [Parity](https://www.parity.io/). You can find more information about the Substrate in our other [post](https://brightinventions.pl/blog/5-benefits-of-substrate-blockchain). 
+
+In Substrate, the business logic of the blockchain is hidden in the runtime. We can easily change it, by using components called pallets. Substrate allows us not only to modify already defined pallets but also to create a new custom one. In this article, we will create a pallet for zk-SNARKs. Thanks to this, we will be able to run a blockchain for the proof verification process. The pallet will allow us to store Circom and SnarkJS artifacts, it will also use a Groth16 for the proof verification. Finally, we will use our zk-SNARK pallet to help Alice join the Bright Coders union.
+
+## Starting point
+
+We have already learned that zk-SNARKs can be used to prove knowledge of a solution for a problem, without revealing it. We just need to provide proof that can be later verified by someone else. Creation of such proof is done in a couple of stages, where we first convert our problem to R1CS (*Rank-1 Constraint System*) form, and then transform it to the QAP (*Quadratic Arithmetic Program*). For this process, we used Circom and SnarkJS. We used a Groth16 as a proving system for assurance of the encryption. Finally, we were able to implement a Rust library, which used artifacts from Circom and SnarkJS, to validate proof.
+
+In the previous post, we show how we can use Circom and SnarkJS in the verification process. Let's remind us what artifacts in the context of those tools do we get:
```

Comment:
"do" do usunięcia --> Let's remind us what artifacts in the context of those tools do we get